### PR TITLE
Rust Alloc and custom Vec

### PR DESCRIPTION
**Pull request by ezrosent**
Currently a work in progress. One thing that's missing is memory measurements in the benchmarks. There is also a problem with `PageAlloc`'s drop implementation (hint: it doesn't have one) that I can fix this evening.

### CODE REVIEW

File: info/custom-vec.md [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+allocation. This interface is different from the standard `malloc`/`free`
+interface from C. Here is a shorter version of the [full
+interface](https://doc.rust-lang.org/nightly/alloc/allocator/trait.Alloc.html):
+
+```rust
+pub unsafe trait Alloc {
+  unsafe fn alloc(&mut self, layout: Layout) -> Result<*mut u8, AllocErr>;
+  unsafe fn dealloc(&mut self, ptr: *mut u8, layout: Layout);
+  // alloc and dealloc are sufficient to implement this trait, but there are
+  // several other methods in the full trait.
```
Comment (joshlf) [inline info/custom-vec.md]: I'd make this "...in the full trait that have default implementations."

File: info/custom-vec.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+}
+```
+
+Where
+[`Layout`](https://doc.rust-lang.org/nightly/alloc/allocator/struct.Layout.html)
+is a type that encapsulates size and alignment information, and
+[`AllocErr`](https://doc.rust-lang.org/nightly/alloc/allocator/enum.AllocErr.html)
+specifies a few common errors that allocators can exhibit (e.g. OOM). The
+biggest difference between this interface and the more traditional one is the
+fact that the caller provides size information during deallocation.
```
Comment (joshlf) [inline info/custom-vec.md]: I'd say that the "biggest differences between this interface and the more traditional one are that allocation requests take an explicit alignment, and size and alignment are provided on deallocation and reallocation."

**Comment by ezrosent** [comment]
Note that while this diff is fairly large, the number that github reports is inflated: that's because Rustfmt is still causing fairly large diffs on older files. I believe the defaults changed at some point. There must be a way to avoid things like this...

File: elfmalloc/src/bin/bench_vec.rs [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+fn format_dur(mut nsecs: f64) -> String {
+    if nsecs >= 1_000_000_000f64 {
+        nsecs /= 1_000_000_000f64;
+        format!("{:.03} s", nsecs)
+    } else if nsecs >= 1_000_000f64 {
+        nsecs /= 1_000_000f64;
+        format!("{:.03} ms", nsecs)
+    } else if nsecs >= 1_000f64 {
+        nsecs /= 1_000f64;
+        format!("{:.03} us", nsecs)
```
Comment (joshlf) [inline elfmalloc/src/bin/bench_vec.rs]: Can you use utf-8 mu?

File: info/custom-vec.md [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+benchmark-n01 bench_push_large_avec_elf               10.529 ms    per iteration
+benchmark-n32 bench_push_large_avec_elf               372.935 ms   per iteration (+/- 1.832 ms)
+benchmark-n01 bench_push_large_vec                    14.107 ms    per iteration
+benchmark-n32 bench_push_large_vec                    556.480 ms   per iteration (+/- 3.350 ms)
+benchmark-n01 bench_push_large_avec_heap              16.099 ms    per iteration
+benchmark-n32 bench_push_large_avec_heap              550.472 ms   per iteration (+/- 3.078 ms)
+```
+
+Here, `elfmalloc` again performs better than the default heap allocator. I
+suspect that this is due to our use of the `BagPipe` data-structure to cache
```
Comment (joshlf) [inline info/custom-vec.md]: BagPipe or Bagpipe? Previously in this doc you use the latter capitalization.

**Pull request merged**